### ISSUE TITLE
getProductUrl does not allow to override the scope in backend context

### ISSUE THREAD

**Issue by sebastian-enz** [magento/magento2#21478]
### Preconditions:
1. Magento instance with several websites
2. Each website has a specific domain (base url)

### Steps to reproduce:
1. Open a product page for a specific website in the admin area
2. Use the $product->getProductUrl() method to retrieve product url

### Expected result:
Returned product url relates to specified website

### Actual result:
Magento always uses the 'admin' scope (and the default base url) to retrieve product url and returns incorrect url.

By calling the method `$product->getProductUrl()` I'm always retreiving the product's url with the admin storekey in it. (for example `http://HOST/admin/catalog/product/view/...`)

I've looked into the problem and the method `getBaseUrl` in `Magento\Framework\Url` offers the possiblity to override the url scope with the parameter `_scope`.

In backend context the subclass `Magento\Backend\Model\Url` is used and the call to [_getScope()](https://github.com/magento/magento2/blob/develop/app/code/Magento/Backend/Model/Url.php#L410) which is executed within the getBaseUrl() method always returns the admin store. `Magento\Framework\Url::_getScope()` uses the data value of `scope` instead.

I think that it should be possible to also override the scope in backend context. (See: https://github.com/magento/magento2/issues/4247#issuecomment-248342560 as well)

Look at issue #4247 as well, which the author decided to close after 3 years of waiting. However, this is STILL an issue, today, in Magento 2.3.x, so I feel like it's appropriate to open a new issue, as the author of the old issue doesn't want to keep his issue open that long.

**Comment by magento-engcom-team** [comment]
Hi @ikk0. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@ikk0 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @shikhamis11 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98467`, `MAGETWO-98468` were created

**Issue Available**: @shikhamis11, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by mohith-ecommerce** [comment]
before saving the product or any attributes of the product setCurrentStore to '0' (zero) you can do this by following these steps

use Magento\Store\Model\StoreManagerInterface;
class YourClassName
{
/**
* @var StoreManagerInterface
*/
protected $_storeManager;
public function __construct(
StoreManagerInterface $storeManager
)
{
$this->_storeManager = $storeManager;
}
public function yourFunctionName()
{
$this->_storeManager->setCurrentStore('0');
//do your work and save
}
It work for me let me know if any issues

**Comment by magento-engcom-team** [comment]
Hi @Nazar65. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if you want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by Nazar65** [comment]
Hi @ikk0 after hour of debugging, i'm able to get frontend url for base website like this
```
        $product->setStoreId(2);
        $url1 = $product->getProductUrl();
        $url = $product->getUrlInStore();
$url = http://magento23dev.vg/index.php/product1.html
$url2 = http://magento23dev.vg/index.php/product1.html
```
But still not able to get Url from second store. working on this now .. 
and @mohith227  solution not work.
